### PR TITLE
[SPARK-28878][SQL] Remove extra project for DSv2 reads with columnar batches

### PR DESCRIPTION
**Pull request by rdblue**
### What changes were proposed in this pull request?

Remove unnecessary physical projection added to ensure rows are `UnsafeRow` when the DSv2 scan is columnar. This is not needed because conversions are automatically added to convert from columnar operators to `UnsafeRow` when the next operator does not support columnar execution.

### Why are the changes needed?

Removes an extra projection and copy.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by prodeezy** [comment]
@rdblue This looks very useful and pertinent to Vectorization support use-case in DSV2 and should help in further improving performance. Thanks for working on this!
Is this mostly pending unit tests/ docs for the "WIP" to come off or is there more work to be one on this optimization?

**Comment by rdblue** [comment]
@prodeezy, I just needed to verify that there are tests for the columnar path and there are: `JavaColumnarDataSourceV2` is used in `DataSourceV2Suite`.

**Comment by rdblue** [comment]
@brkyvz, can you take a look at this?

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala [hunk 1]
```diff
@@ -140,13 +140,18 @@ object DataSourceV2Strategy extends Strategy with PredicateHelper {
...
-      val plan = BatchScanExec(output, scan)
+      val batchExec = BatchScanExec(output, scan)
 
       val filterCondition = postScanFilters.reduceLeftOption(And)
-      val withFilter = filterCondition.map(FilterExec(_, plan)).getOrElse(plan)
+      val withFilter = filterCondition.map(FilterExec(_, batchExec)).getOrElse(batchExec)
 
-      // always add the projection, which will produce unsafe rows required by some operators
-      ProjectExec(project, withFilter) :: Nil
+      val withProjection = if (withFilter.output != project || !batchExec.supportsColumnar) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala]: BTW, shall we do the same for streaming scan?

Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala]: I'm not as familiar with the streaming side. If we support columnar reads there, we probably can.

**Comment by rdblue** [comment]
Thanks for merging this, @cloud-fan!

**Pull request closed without merging**